### ISSUE TITLE
Validate PyPi based images for central services

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#11354]
**Impact of the new feature**
WMCore central services

**Is your feature request related to a problem? Please describe.**
This is part of the effort to move away from RPM-based images, and fully adopt the PyPi packaging for WMCore deployment.

**Describe the solution you'd like**
This issue is a subtask of https://github.com/dmwm/WMCore/issues/10178, just so we can distribute work between implementation of the Pypi-based dockerfile and its basic validation, before we can actually push upstream the new dockerfiles.

Now that we have the very first version of Dockerfile and images for WMCore services, based on the PyPi packages, see:
https://github.com/dmwm/WMCore/issues/11310#issuecomment-1293404825

we can start running some basic tests for these new images. For that, we need to:
* wait for Apache to be deployed in the testX clusters (to get CouchDB replication working with the agents)
* deploy CouchDB (it can be either v2 (CMS RPM-based) or v4 (vanilla upstream CouchDB))
* deploy one of the images uploaded to the CERN registry (see Valentin's update above). It could be ReqMgr2

The basic tests involve:
* starting the service
* stopping the service
* query one simple API, e.g.: `reqmgr2/data/info` 

@vkuznet can update this GH issue with further details.

**Describe alternatives you've considered**
None

**Additional context**
None

**Comment by vkuznet** [comment]
This issue depends on availability of apache FE (as daemonset) on dev clusters. Here is relevant [CMSKUBERNETES-200](https://its.cern.ch/jira/browse/CMSKUBERNETES-200) ticket.

**Comment by vkuznet** [comment]
I tried to install `reqmgr2` image on dev cluster and immediately found several issues. I list them here for clarify:
- images require `rotatelogs`, therefore I added `apache2-utils` package via `apt-get install` to provide it
- since we use `run.sh` script in our manifest files we need to provide this for every image
- the `run.sh` script from RPM based images relies on `manage` from our deployment which no longer exists in pypi based installation. Therefore I re-wrote `run.sh` script as simple as following:
```
#!/bin/bash
# script to start ReqMgr2

STATEDIR=/data/state
LOGDIR=/data/logs
CFGFILE=/etc/secrets/config.py
mkdir -p /data/{logs,state}
wmc-httpd -r -d $STATEDIR -l "|rotatelogs $LOGDIR/reqmgr2-%Y%m%d-`hostname -s`.log 86400" $CFGFILE
```
This is bare minimum we need to start the service
- I rebuilt `reqmgr2` image with these changes
- i was able to deploy new image and run it, the `reqmgr2` was started, but I was not yet tested its functionality.
- I uploaded changes to CMSKubernetes repo to all pypi images
- New images are build for all WMCore services and uploaded to registry.cern.ch with pypi tag

**Comment by vkuznet** [comment]
After many iterations I finally came up with images which work while I deploy them on dev cluster. In short, the current DMWM images are based on RPM based deployment which requires setting up `/data/srv` area, having `manage` script, etc. The pypi images do not have that. Therefore, I was forced to adjust corresponding `run.sh` scripts in these images to imitate RPM based deployment. In other words, I manually created `/data/srv` area, and put relevant files in appropriate places. As a result I produced `run.sh` script which starts WMCore service based on `wmc-httpd` daemon (this is what is used in RPM based `manage` script). You may have a look at corresponding [pypi/reqmgr2/run.sh](https://github.com/dmwm/CMSKubernetes/blob/master/docker/pypi/reqmgr2/run.sh) script from reqmgr2. You'll find it similar to already existing [reqmgr2/run.sh](https://github.com/dmwm/CMSKubernetes/blob/master/docker/reqmgr2/run.sh) and if you wish you can make a diff.

After these changes (which I applied to all WMCore services in CMSKubernetes/docker/pypi area) I rebuilt all images and uploaded them to registry.cern.ch

After that, I was able to deploy pypi based reqmgr2 service on test7 cluster and it runs fine, see its log:
```
k logs -n dmwm reqmgr2-5b8b7b9ccc-krcr7
INFO Files in /etc/grid-security
total 24K
drwxr-xr-x. 1 root root system_u:object_r:var_lib_t:s0 4.0K Nov  4 17:12 .
drwxr-xr-x. 1 root root system_u:object_r:var_lib_t:s0  107 Nov  4 17:12 ..
-rw-r--r--. 1 root root system_u:object_r:var_lib_t:s0 2.1K Nov  4 17:12 cmsweb-test7-z3qbetkyetbe-node-2.cern.ch.grid.crt
-rw-------. 1 root root system_u:object_r:var_lib_t:s0 1.7K Nov  4 17:12 cmsweb-test7-z3qbetkyetbe-node-2.cern.ch.grid.key
-rw-r--r--. 1 root root system_u:object_r:var_lib_t:s0 2.9K Nov  4 17:12 cmsweb-test7-z3qbetkyetbe-node-2.cern.ch.grid.pem
-rw-r--r--. 1 root root system_u:object_r:var_lib_t:s0 2.9K Nov  4 17:12 hostcert.pem
-rw-------. 1 root root system_u:object_r:var_lib_t:s0 1.7K Nov  4 17:12 hostkey.pem
reqmgr2 not running, not killing
[04/Nov/2022:17:12:16]  Creating CouchDB connection instances to CouchDB host: 'https://cmsweb-test7.cern.ch/couchdb' ...
[04/Nov/2022:17:12:16]  Creating CouchDB connection to 'reqmgr_workload_cache' ...
[04/Nov/2022:17:12:16]  Creating CouchDB connection to 'reqmgr_auxiliary' ...
[04/Nov/2022:17:12:16]  INFO: loading ui into /reqmgr2
[04/Nov/2022:17:12:16] TemplatedPage ### ReqMgr uses JINJA templates
[04/Nov/2022:17:12:16] TemplatedPage Templates are located in: /data/apps/reqmgr2/data/html/ReqMgr/templates
[04/Nov/2022:17:12:16]  INFO: starting server in /data/srv/state/reqmgr2
[04/Nov/2022:17:12:16] ENGINE Bus STARTING
[04/Nov/2022:17:12:16] ENGINE Serving on http://0.0.0.0:8246
[04/Nov/2022:17:12:16] ENGINE Bus STARTED
[04/Nov/2022:17:14:19]  ERROR: authz denied, front-end headers not present
[04/Nov/2022:17:14:19] reqmgr2-5b8b7b9ccc-krcr7 10.100.158.192 "GET /reqmgr2/data/info HTTP/1.1" 403 Forbidden [data: 98 in 750 out 1413 us ] [auth: - "" "" ] [ref: "" "curl/7.29.0" ]
```

But we still can't access service from cmsweb-test7 URL since our apache FE rules are not valid on dev cluster. Below I place appropriate set of actions to @muhammadimranfarooqi and @arooshap . Meanwhile, I only was able to place curl call from another pod and it was semi-successful, i.e. the request came through but because it was done from the pod and not from apache it fails to provide appropriate HTTP header.

To sum-up, the deployment and service are fine, but we need to fix apache FE. 

@muhammadimranfarooqi , @arooshap I found that apache FE rules on test7 clusters are not correct. Here is what we have:
```
[_frontend@cmsweb-test7-z3qbetkyetbe-node-1 data]$ grep reqmgr2 /data/srv/current/config/frontend/backends.txt
^/auth/complete/reqmgr2(?:/|$) vocms0132.cern.ch|vocms0731.cern.ch :affinity
```
and, we need rule to point to local `http://reqmgr2.dmwm:8246`. So, I would assume that `backends.txt` files supplied to apache FE is wrong for WMCore dev-clusters. Please fix it and redeploy apache FE on all WMCore dev clusters.

**Comment by vkuznet** [comment]
Now, the apache FE rules are fixed and we get the proper response from reqmgr2 services:
```
scurl https://cmsweb-test7.cern.ch/reqmgr2/data/info
{"result": [
 {
  "wmcore_reqmgr_version": "2.1.4.pre3",
  "reqmgr_db_info": {
,,,
```

@amaltaro , please have a look at reqmgr2 on test7 cluster. Now it is deployed with pypi based image and in my view behave normally.

**Comment by amaltaro** [comment]
@vkuznet Valentin, if I deploy these Pypi-based images with the CMSKubernetes `./scripts/deploy-srv.sh` script, what is the tag that I should use to pull the correct images?
Do I also have to replace the services manifest files as you did in your post above?

**Comment by vkuznet** [comment]
```
./scripts/deploy-srv.sh reqmgr2 pypi
```
So far, I only used image tag pypi (where I used latest WMCore tag at a time). Once we try them out we can decide on tag naming convention, e.g. `pypi-<actual_wmcore_tag>`. Feel free to propose specific naming convention though and I can generate images.

**Comment by amaltaro** [comment]
Quick update on this: I managed to deploy some of the Pypi-based services, like: reqmgr2-*, reqmon-* and workqueue.

However, all of the microservices fail to start and I think the reason is that we are trying to run them with an incorrect/inexistent service configuration file. This CMSKubernetes should fix it:
https://github.com/dmwm/CMSKubernetes/pull/1253

but I think a new docker image needs to be created. Can you please take care of this, @vkuznet ?

**Comment by amaltaro** [comment]
@vkuznet Valentin, thank you for creating those images and providing further instructions on creating/uploading the images.

Unfortunately, things are still fairly broken after deploying all the WMCore images with the following changes made to my clone of CMSKubernetes repo, also provided in: https://github.com/dmwm/CMSKubernetes/pull/1258

and deploying WMCore services with:
```
serviceList="reqmgr2 \
             reqmgr2-tasks \
             reqmgr2ms-monitor \
             reqmgr2ms-transferor \
             reqmgr2ms-rulecleaner \
             reqmgr2ms-output \
             reqmgr2ms-unmerged-t1 \
             reqmgr2ms-unmerged-t2t3us \
             reqmgr2ms-unmerged-t2t3 \
             workqueue \
             reqmon \
             reqmon-tasks"
### now deploy these services
for service in $serviceList
do
  ./scripts/deploy-srv.sh $service pypi-20221119
done
```

Here is the current status of my pods (in test9):
```
amaltaro@lxplus8s20:~/private/cmsweb-deploy/CMSKubernetes $ kubepods dmwm
NAME                               READY   STATUS             RESTARTS        AGE
ms-monitor-6fb5f67f4c-jmm2p        0/1     CrashLoopBackOff   7 (2m20s ago)   14m
ms-output-5bf65b68bc-xzf8v         0/1     Init:0/1           0               14m
ms-rulecleaner-5fb7459d58-jc8g2    0/1     CrashLoopBackOff   7 (2m10s ago)   14m
ms-transferor-7c8f7b68bd-77snz     0/1     CrashLoopBackOff   7 (2m10s ago)   14m
ms-unmer-t1-6c8d795d-ncxbj         1/1     Running            0               14m
ms-unmer-t2t3-68c8d5bf66-x8q8q     1/1     Running            0               14m
ms-unmer-t2t3us-74b959b57d-8pkms   1/1     Running            1 (13m ago)     14m
reqmgr2-7d6bdd9b76-njplk           0/1     ImagePullBackOff   0               15m
reqmgr2-tasks-5d84779fbb-nbl26     0/1     ImagePullBackOff   0               14m
reqmon-77f7fc45d5-4l667            1/1     Running            0               14m
reqmon-tasks-668c7c8594-97dk9      1/1     Running            0               14m
workqueue-75c97576f-l95sf          1/1     Running            0               14m
```

where `reqmgr2` seems to be failing because you did not create the required `pypi-20221119`, and the microservices seem to be facing exactly the same problem as before:
```
amaltaro@lxplus8s20:~/private/cmsweb-deploy/CMSKubernetes/kubernetes/cmsweb $ kubectl logs ms-transferor-7c8f7b68bd-77snz -n dmwm
/usr/local/bin/wmc-httpd: /etc/secrets/config.py: invalid configuration file
```

Once you resolve all these issues and you can confirm that ALL the WMCore services can be properly deployed, I can resume testing this.

**Comment by vkuznet** [comment]
Alan, the pypi-20221119 images exists, e.g. please see https://registry.cern.ch/harbor/projects/1771/repositories/reqmgr2ms-monitor/artifacts-tab/artifacts/sha256:92d00b1d3dc8a981c60065693d597bee9758d2483837e175f646fc40aebe39be
And, you can verify this by simply using docker pull, e.g.
```

vocms0181(21:40:28) docker > docker pull registry.cern.ch/cmsweb/reqmgr2ms-transferor:pypi-20221119
pypi-20221119: Pulling from cmsweb/reqmgr2ms-transferor
f606d8928ed3: Already exists
47db815c6a45: Already exists
bf4849400000: Already exists
a572f7a256d3: Already exists
8f7d05258955: Already exists
7110f04115ae: Already exists
e06ef06dbaac: Already exists
da646a36a161: Already exists
345557d01003: Already exists
cd554b9e3011: Already exists
f093f31c47c6: Already exists
bd45297915fa: Already exists
ff349b1a1fc5: Pull complete
52c7bb2cf594: Pull complete
351b1e13d08e: Pull complete
dee7f81cbdbe: Pull complete
649692a2fcc8: Pull complete
84166ae8bbe8: Pull complete
Digest: sha256:beeb8eaecf5ff34460d7052458d2d94bf18ef1bd560b05639bdbd4d660d6082c
Status: Downloaded newer image for registry.cern.ch/cmsweb/reqmgr2ms-transferor:pypi-20221119
registry.cern.ch/cmsweb/reqmgr2ms-transferor:pypi-20221119
vocms0181(21:40:41) docker > docker images
REPOSITORY                                     TAG             IMAGE ID       CREATED        SIZE
registry.cern.ch/cmsweb/reqmgr2ms-transferor   pypi-20221119   5e208bb08c93   2 days ago     1.18GB
```

To investigate the actual issue please provide cluster name you are using, and I can have a look.

**Comment by amaltaro** [comment]
Please read again the details I provided in my previous post. But if you want to look into the pods, this is test9 cluster.

**Comment by vkuznet** [comment]
Alan, I finally got to the root of the problem. The `scripts/deploy-srv.sh` does the following
```
curl -ksLO https://raw.githubusercontent.com/dmwm/CMSKubernetes/master/kubernetes/cmsweb/services/$srv.yaml
```
see https://github.com/dmwm/CMSKubernetes/blob/master/kubernetes/cmsweb/scripts/deploy-srv.sh#L155

Therefore, when you run it you'll get `reqmgr2ms` images instead of `reqmgr2ms-transferor` (or any other service) since this is what is available in main repository, see https://github.com/dmwm/CMSKubernetes/blob/master/kubernetes/cmsweb/services/reqmgr2ms-transferor.yaml#L89

Therefore, in order to properly deploy new images you have two options:
- either clone and patch CMSKubernetes repo and adjust your `deploy-srv.sh` to pick up from your clone, or
- instead of using `scripts/deploy-srv.sh` you should just use plain k8s commands, e.g.
```
# copy your manifest file, e.g. reqmgr2ms-transferor.yaml to your location, e..g
cp services/reqmgr2ms-transferor.yaml /path/reqmgr2ms-transferor.yaml

# adjust your manifest file to use proper imagetag

# deploy it with kubectl command
kubectl apply --validate=false -f /pathreqmgr2ms-transferor.yaml
```
For your convenience here is a diff I made for `reqmgr2ms-transferor.yaml` when I was debugging this issue:
```
diff -u services/reqmgr2ms-transferor.yaml reqmgr2ms-transferor.yaml
--- services/reqmgr2ms-transferor.yaml  2022-11-01 12:54:35.000000001 +0100
+++ reqmgr2ms-transferor.yaml   2022-11-21 23:12:28.000000001 +0100
@@ -54,6 +54,7 @@
     sudo cp /host/etc/grid-security/* /etc/grid-security
     echo 'INFO Files in /etc/grid-security'
     ls -lahZ /etc/grid-security
+    cat /data/run.sh
     #su  -c "cd /data && /data/run.sh" --preserve-environment _reqmgr2ms
     cd /data && /data/run.sh
 ---
@@ -86,7 +87,9 @@
         runAsGroup: 1000
         fsGroup: 2000
       containers:
-      - image: registry.cern.ch/cmsweb/reqmgr2ms #imagetag
+#       - image: registry.cern.ch/cmsweb/reqmgr2ms #imagetag
+      - image: registry.cern.ch/cmsweb/reqmgr2ms-transferor:pypi-20221121
+        imagePullPolicy: Always
         name: ms-transferor
         lifecycle:
           postStart:
```
And, to avoid mistakes I created new `pypi-20221121` tag.

Now, I deployed `reqmgr2ms-transferor` (using `kubectl apply` method) on test9 cluster and things looks completely different. Now it complains about the following:
```
k logs -n dmwm ms-transferor-5cb4f65fb4-f5gc7
...
ms-transferor not running, not killing
    BaseRotatingHandler.__init__(self, filename, 'a', encoding, delay)
  File "/usr/local/lib/python3.8/logging/handlers.py", line 55, in __init__
    logging.FileHandler.__init__(self, filename, mode, encoding, delay)
  File "/usr/local/lib/python3.8/logging/__init__.py", line 1147, in __init__
    StreamHandler.__init__(self, self._open())
  File "/usr/local/lib/python3.8/logging/__init__.py", line 1176, in _open
    return open(self.baseFilename, self.mode, encoding=self.encoding)
FileNotFoundError: [Errno 2] No such file or directory: '/logs/reqmgr2ms/heartbeatMonitor_transferor-ms-transferor-5cb4f65fb4-f5gc7-20221121.log'

[21/Nov/2022:22:12:50]  WATCHDOG: server exited with exit code 1
```
which still implies that configuration file `config-transferor.py` is not perfect. In particular, it fails due to this line in configuration file:
```
heartbeatMonitor.log_file = '%s/logs/reqmgr2ms/heartbeatMonitor_transferor-%s-%s.log' % (__file__.rsplit('/', 4)[0],
                                                                                         HOST.split('.', 1)[0],
                                                                                         time.strftime("%Y%m%d"))
```
which relies on the old RPM based directory stricture, i.e. the config file should be located in `/data/srv/current/config/reqmgr2ms-transferor` area. Therefore, I suggest that you revisit configuration file and adjust it according to new location (now we read configs from /etc/secrets and not from /data/srv/current/config area).

**Comment by amaltaro** [comment]
> And, to avoid mistakes I created new pypi-20221121 tag.

@vkuznet Valentin, I can confirm that tweaking `deploy-srv.sh` script to use the yaml files from the local node (instead of downloading it with curl) does the trick and I have all the pods in `Running` state now.

However, I think most - or maybe all - of the microservices crashed during the application startup, as you mentioned in your previous posts. Let me list here a few problems that I could spot:
1) you are loading the configuration file from a wrong location: https://github.com/dmwm/CMSKubernetes/blob/master/docker/pypi/reqmgr2ms-unmerged/run.sh#L7, which likely explains why the service crashes with:
```
FileNotFoundError: [Errno 2] No such file or directory: '/logs/reqmgr2ms/heartbeatMonitor_transferor-ms-transferor-5b55cfd544-686hl-20221122.log'
```
note that it's missing the `/data/srv` prefix.

2) the configuration directory (/data/srv/current/config) is empty
```
_reqmgr2ms@ms-transferor-5b55cfd544-686hl:/data/srv/current$ ls config/reqmgr2ms/
_reqmgr2ms@ms-transferor-5b55cfd544-686hl:/data/srv/current$
_reqmgr2ms@ms-transferor-5b55cfd544-686hl:/data/srv/current$ ls /etc/secrets/
ReqMgr2MSSecrets.py  config-transferor.py
_reqmgr2ms@ms-transferor-5b55cfd544-686hl:/data/srv/current$ ls auth/
proxy  reqmgr2ms  wmcore-auth
```

3) the secrets area is messed up (the service configuration should not be here!)
```
_reqmgr2ms@ms-transferor-5b55cfd544-686hl:/data/srv/current$ ls auth/reqmgr2ms/
ReqMgr2MSSecrets.py  config-transferor.py  dmwm-service-cert.pem  dmwm-service-key.pem	header-auth-key
```

In short, all these problems seem to arise from the new `run.sh` script that you wrote here:
https://github.com/dmwm/CMSKubernetes/blob/master/docker/pypi/reqmgr2ms-unmerged/run.sh#L7

Before I can resume validating the WMCore images based on Pypi, I would like you to:
* fix the images and its embedded scripts
* make sure that all services can be deployed (already the case)
* make sure that all the pods are in a `Running` state (already the case, except for MSRuleCleaner - maybe unrelated issue with mongodb)
* make sure that all the services are actully running INSIDE the pods (no crashes in the component, having a PID for them)

Once you can confirm that the following services are deployable, container running and application running:
```
ms-monitor
ms-output
ms-rulecleaner
ms-transferor
ms-unmer-t1
ms-unmer-t2t3
ms-unmer-t2t3us
reqmgr2
reqmgr2-tasks
reqmon
reqmon-tasks
workqueue
```

I can come back to this. If you need my input on any of this, please let me know. But I think you have all the information and tools to resolve it by yourself.

**Comment by vkuznet** [comment]
Alan, of course I know and understand all of this. The main issue here is that we apply RPM based procedure to pypi, i.e. all configuration files, directory structure are from RPM based deployment and it is not what we have with pypi. So, the question here is do we need to stick RPM based deployment configs, directory structure or not? If we decide to stick with RPM based layout then we need changes to `run.sh` to recreate all of these, e.g. `/data/srv` is not necessary for pypi deployment, but all configs rely on this. Otherwise, if we decide that service can run with provided configs and run anywhere we need changes to config files. We are making a big change here by moving away from RPM and it may affect how we deploy services and how we structure configuration. Please provide your feedback on this conceptual aspect.

**Comment by amaltaro** [comment]
I think you are raising this point a bit late in the chain. This should have been discussed before you started making all the Dockerfile changes and asking for further validation of the images. If it's a design-like question, it definitely needs to be discussed before the implementation is made.

Having said that, my opinion on this is that we should keep the same directory structure as we have right now. Unless it requires a tremendous effort or there is a strong reason not to do so. In other words, we should try to stick to the same organization that we have right now, such that:
* `/data/srv/state` is used for data that is meant to be persisted between deployments (actual data, databases, cache, etc)
* `/data/srv/logs` has to contain the service logs
* `/data/srv/current/config` contains the service configuration file(s)
* `/data/srv/current/auth` contains any secrets and additional configuration files
* `manage` script, or something similar that we can start/stop/status/restart the service from within the container

In addition to this, I noticed that the reqmgr2 image does not have any text editor tool (no vi, vim, emacs, nano, not even less). This should definitely be corrected.

**Comment by vkuznet** [comment]
Alan, I raised many design questions, and so far I never see them addressed. Said that, it is a fact that pip installation is not CMS RPM based one, it is clear. In fact, I already raised this design question a first time I adopted RPM based installation to images, and I got the impression that we need to preserve anything. Said that, I'll correct `run.sh` to preserve directory structure and create new images with them. In my opinion it is wrong since it forces to always have this directory structure even in a dev environment, e.g. on my laptop, and in certain situation dev environment may not have ability to set `/data` structure since I may not have proper admin rights.

Regarding vi and other tools. I intentionally do not install them to avoid blowing up image size. If you ned them in a pod just install them there since you have sudo access. Here is how I usually get vi in a pod which does not have it:
```
# step 1: login to the pod

# step 2: perform apt-get update
sudo apt-get update

# search for necessary tool
sudo apt-cache search vim

# install a tool you want
sudo apt-get install vim
```
This has benefit that within your pod you can install tools you need and after that you may restart pod and all tools disappear since they are not part of the image. This is very good benefit in terms of security and reduction of image size.

**Comment by vkuznet** [comment]
Alan, I adjusted `run.sh` to imitate RPM based dir structure and I am able to start services, but I face another problem, e.g. in ms-monitor and ms-transferor I see this error:
```
==> /data/srv/logs/reqmgr2ms/reqmgr2ms-20221122-ms-monitor-55d457b49b-wxmdn.log <==
  File "/usr/local/lib/python3.8/site-packages/rucio/common/config.py", line 117, in config_get_bool
    return get_config().getboolean(section, option)
  File "/usr/local/lib/python3.8/site-packages/rucio/common/config.py", line 217, in get_config
    __CONFIG = Config()
  File "/usr/local/lib/python3.8/site-packages/rucio/common/config.py", line 241, in __init__
    raise RuntimeError('Could not load Rucio configuration file. '
RuntimeError: Could not load Rucio configuration file. Rucio looked in the following paths for a configuration file, in order:
        /opt/rucio/etc/rucio.cfg

[22/Nov/2022:23:02:18]  WATCHDOG: server exited with exit code 1
```
which tells me that both of these images need rucio config. My question is where I can find it and how it should be supplied to MS services? Again, I have no idea because I never deployed and worked with this services. Please advise.

**Comment by amaltaro** [comment]
@vkuznet I do not have a precise answer on this, but I can give you a couple of pointers that might help you out.

I just happened to work on the rucio spec file and I saw this tweak to keep the CMS rucio config:
https://github.com/cms-sw/cmsdist/blob/comp_gcc630/py3-rucio-clients.spec#L11

but I understand that this is no longer relevant for PyPi packaging.

Looking into how we use/used to deploy the microservices RPM, we have some special tweak for the rucio.cfg according to where we are actually going deploy/run that service (cmsweb / testbed / dev):
https://github.com/dmwm/deployment/blob/master/reqmgr2ms/deploy#L68

so we might have to plug in a similar logic to the run.sh script (maybe we push the rucio.cfg to the services_config repository and curl/wget it during startup)?

**Comment by vkuznet** [comment]
@amaltaro , I managed to get all DMWM pods starts correctly and majority are running fine according to logs, e.g.
```
k logs -n dmwm workqueue-6b6f959659-rsb6j
...
[23/Nov/2022:15:02:33]  INFO: starting server in /data/srv/state/workqueue
[23/Nov/2022:15:02:33] ENGINE Bus STARTING
[23/Nov/2022:15:02:33] ENGINE Serving on http://0.0.0.0:8240
[23/Nov/2022:15:02:33] ENGINE Bus STARTED
```
or
```
k logs -n dmwm ms-transferor-85b77fb95d-qlb2g
...
2022-11-23 14:37:13,786:INFO:_cplogging: [23/Nov/2022:14:37:13] ENGINE Serving on http://0.0.0.0:8247
2022-11-23 14:37:13,787:INFO:HeartbeatMonitorBase: Checking Thread status...
[23/Nov/2022:14:37:13] ENGINE Bus STARTED
2022-11-23 14:37:13,788:INFO:_cplogging: [23/Nov/2022:14:37:13] ENGINE Bus STARTED
2022-11-23 14:37:14,053:INFO:HeartbeatMonitorBase: Uploaded to WMStats...

==> /data/srv/logs/reqmgr2ms/heartbeatMonitor_transferor-ms-transferor-85b77fb95d-qlb2g-20221123.log <==
2022-11-23 14:47:14,275:INFO:HeartbeatMonitorBase:Checking Thread status...

==> /data/srv/logs/reqmgr2ms/reqmgr2ms-20221123-ms-transferor-85b77fb95d-qlb2g.log <==
2022-11-23 14:47:14,275:INFO:HeartbeatMonitorBase: Checking Thread status...

==> /data/srv/logs/reqmgr2ms/heartbeatMonitor_transferor-ms-transferor-85b77fb95d-qlb2g-20221123.log <==
2022-11-23 14:47:14,519:INFO:HeartbeatMonitorBase:Uploaded to WMStats...

==> /data/srv/logs/reqmgr2ms/reqmgr2ms-20221123-ms-transferor-85b77fb95d-qlb2g.log <==
2022-11-23 14:47:14,519:INFO:HeartbeatMonitorBase: Uploaded to WMStats...

==> /data/srv/logs/reqmgr2ms/heartbeatMonitor_transferor-ms-transferor-85b77fb95d-qlb2g-20221123.log <==
2022-11-23 14:57:14,727:INFO:HeartbeatMonitorBase:Checking Thread status...
2022-11-23 14:57:14,976:INFO:HeartbeatMonitorBase:Uploaded to WMStats..
```
But there are three unmerged services which shows identical error in their logs:
```
k logs -n dmwm ms-unmer-t1-59cc6b747d-6chqb
...
==> /data/srv/logs/reqmgr2ms/reqmgr2ms-20221123-ms-unmer-t1-59cc6b747d-6chqb.log <==
2022-11-23 15:02:08,913:ERROR:CherryPyPeriodicTask:    File "/usr/local/lib/python3.8/site-packages/WMCore/Services/Requests.py", line 173, in makeRequest
2022-11-23 15:02:08,913:ERROR:CherryPyPeriodicTask:      result, response = self.makeRequest_pycurl(uri, data, verb, headers)
2022-11-23 15:02:08,913:ERROR:CherryPyPeriodicTask:    File "/usr/local/lib/python3.8/site-packages/WMCore/Services/Requests.py", line 190, in makeRequest_pycurl
2022-11-23 15:02:08,913:ERROR:CherryPyPeriodicTask:      response, result = self.reqmgr.request(uri, data, headers, verb=verb,
2022-11-23 15:02:08,913:ERROR:CherryPyPeriodicTask:    File "/usr/local/lib/python3.8/site-packages/Utils/PortForward.py", line 67, in portMangle
2022-11-23 15:02:08,913:ERROR:CherryPyPeriodicTask:      return callFunc(callObj, newUrl, *args, **kwargs)
2022-11-23 15:02:08,914:ERROR:CherryPyPeriodicTask:    File "/usr/local/lib/python3.8/site-packages/WMCore/Services/pycurl_manager.py", line 317, in request
2022-11-23 15:02:08,914:ERROR:CherryPyPeriodicTask:      curl.perform()
2022-11-23 15:02:08,914:ERROR:CherryPyPeriodicTask:  pycurl.error: (60, 'SSL certificate problem: self signed certificate in certificate chain')
2022-11-23 15:02:08,952:ERROR:LogDB: LogDBBackend post API failed, error=(60, 'SSL certificate problem: self signed certificate in certificate chain')
```
Since log does not show me which certificate it tries to use I have no idea what it is and why it is needed. I hope you can have a look and provide your feedback. The pods have their certificates in /etc/grid-security, e.g.
```
k logs -n dmwm ms-unmer-t1-59cc6b747d-6chqb
INFO Files in /etc/grid-security
total 88K
drwxr-xr-x. 1 root root system_u:object_r:var_lib_t:s0 4.0K Nov 23 15:02 .
drwxr-xr-x. 1 root root system_u:object_r:var_lib_t:s0  107 Nov 23 15:02 ..
drwxr-xr-x. 2 root root system_u:object_r:var_lib_t:s0  40K Nov 23 13:45 certificates
-rw-r--r--. 1 root root system_u:object_r:var_lib_t:s0 2.1K Nov 23 15:02 cmsweb-test9-646fxrusxjd5-node-2.cern.ch.grid.crt
-rw-------. 1 root root system_u:object_r:var_lib_t:s0 1.7K Nov 23 15:02 cmsweb-test9-646fxrusxjd5-node-2.cern.ch.grid.key
-rw-r--r--. 1 root root system_u:object_r:var_lib_t:s0 2.9K Nov 23 15:02 cmsweb-test9-646fxrusxjd5-node-2.cern.ch.grid.pem
-rw-r--r--. 1 root root system_u:object_r:var_lib_t:s0 2.9K Nov 23 15:02 hostcert.pem
-rw-------. 1 root root system_u:object_r:var_lib_t:s0 1.7K Nov 23 15:02 hostkey.pem
```
but I'm not sure if it is the ones which shows in log. At the end I think it is only issue with certificate itself rather then image/services.

**Comment by amaltaro** [comment]
> Due to this change I can't commit my changes as is since they will break current (RPM based image deployment). Therefore, we have two options:

Valentin, instead of replicating Dockerfiles in the repository, I would rather stick to only one and make changes not backwards compatible.

Hence, I would suggest you to create a PR renaming the `reqmgr2ms` images to the new PyPi based images, e.g. `reqmgr2ms-output` and we test it properly before actually merging it.

If you prefer to go with another temporary place for those, that is fine with me as well. But we need to make sure that the temporary does not become permanent long-after we are settled with PyPi.

PS.: I still have to read a few others points that you made above.

**Comment by vkuznet** [comment]
Here is corresponding ticket for HTTP group to allow migration from RPM to pypi based images: https://its.cern.ch/jira/browse/CMSKUBERNETES-208

**Comment by vkuznet** [comment]
And, here is PR with changes to k8s manifest files required to perform the transition: https://github.com/dmwm/CMSKubernetes/pull/1261

**Comment by vkuznet** [comment]
yes, I suggest to close this issue as completed.
